mutagen/optimfrog.py

```diff
@@ -41,25 +53,33 @@ class OptimFROGInfo(StreamInfo):
         channels (`int`): number of audio channels
         length (`float`): file length in seconds, as a float
         sample_rate (`int`): audio sampling rate in Hz
+        bits_per_sample (`int`): the audio sample size
+        version (int): encoder version
     """
 
     @convert_error(IOError, OptimFROGHeaderError)
     def __init__(self, fileobj):
         """Raises OptimFROGHeaderError"""
 
         header = fileobj.read(76)
-        if (len(header) != 76 or not header.startswith(b"OFR ") or
-                struct.unpack("<I", header[4:8])[0] not in [12, 15]):
+        if len(header) != 76 or not header.startswith(b"OFR "):
+            raise OptimFROGHeaderError("not an OptimFROG file")
+        data_size = struct.unpack("<I", header[4:8])[0]
+        if data_size != 12 and data_size < 15:
             raise OptimFROGHeaderError("not an OptimFROG file")
         (total_samples, total_samples_high, sample_type, self.channels,
          self.sample_rate) = struct.unpack("<IHBBI", header[8:20])
         total_samples += total_samples_high << 32
         self.channels += 1
+        self.bits_per_sample = SAMPLE_TYPE_BITS.get(sample_type)
         if self.sample_rate:
             self.length = float(total_samples) / (self.channels *
                                                   self.sample_rate)
         else:
             self.length = 0.0
+        if data_size >= 15:
```

Comment:
When is data_size < 15? Which would result in self.version not existing.

Do we need self.version? MP3 has a encoder_info string, maybe that's better?

Reply:
> When is data_size < 15? Which would result in self.version not existing.

There are basically three variants, based on the OptimFROG version:

- 12 bytes header
- 15 bytes header (not quite sure which version added these 3 bytes, but likely very early)
- >= 15 bytes with OptimFROG >= 5.100

> Do we need self.version? MP3 has a encoder_info string, maybe that's better?

I think it is useful. I based this on Musepack which has a `version`. `encoder_info` sounds pretty generic, what exactly is stored there? What about with using `encoder_version` here?

Reply:
After looking at the MP3 encoder_info I think this is good, I will change this for OptimFROG and TAK

Reply:
I might be overly worried regarding future changes here... ymmv

Reply:
> I might be overly worried regarding future changes here... ymmv

This definitely was an issue in the old code, because it checked strictly for 12 or 15 bytes. So all changes to the header length would cause the code to need to get updated.

But the latest spec is this:

```
    4 bytes 'OFR '
    4 bytes main size = S >= 15
    6 bytes total number of samples in the audio file
    1 byte  sample type (see appendix A)
    1 byte  channel configuration (see appendix B)
    4 bytes sample rate
    2 bytes encoder ID (see appendix C)
    1 byte  compression (see appendix D)
    X bytes optional additional fields, where X = S - 15
        2 bytes optional minDecoderVersion - 4500
```

The header now has at least 15 bytes, but is extensible. That means we can keep the 12 bytes test around for backwards compatibility with very old files, but for all other cases testing for >= 15 bytes should be forward compatible.

Reply:
Changed to `encoder_info` in https://github.com/quodlibet/mutagen/pull/406/commits/87aa03f5d877983012490401e083bdadafccb8fe